--- opencdk-0.5.3/src/opencdk.m4.orig	2002-11-12 22:50:51.000000000 +0100
+++ opencdk-0.5.3/src/opencdk.m4	2003-12-18 20:45:26.876176360 +0100
@@ -10,7 +10,7 @@
 dnl AM_PATH_LIBOPENCDK([MINIMUM-VERSION, [ACTION-IF-FOUND [, ACTION-IF-NOT-FOUND ]]])
 dnl Test for libopencdk, and define CDK_CFLAGS and CDK_LIBS
 dnl
-AC_DEFUN(AM_PATH_LIBOPENCDK,
+AC_DEFUN([AM_PATH_LIBOPENCDK],
 [dnl
 dnl Get the cflags and libraries from the opencdk-config script
 dnl
